### PR TITLE
rewrite vscode-deno

### LINKED ISSUE #65: Automatic semicolon insertion on format code.
**Describe the bug**

Semicolons are automatically inserted when selecting 'Format Document' `Shift+Alt+F'. The plugin does not seem to be respecting existing formatting preferences.

**To Reproduce**

1. Ensure automatic semicolons are disabled.
2. Install the plugin
3. Format TypeScript document (note semicolons are being inserted)
4. Uninstall the plugin
5. Format TypeScript document (note semicolons are not being inserted)

**Expected behavior**

It is assumed that the plugin should only provide declarations for Deno's globalThis environment as well as allowing for module resolution via URL. All other user preferences should be left as it, or inherited from the users current TypeScript preferences.

### Versions

deno 0.30.0
v8 8.1.108
typescript 3.7.2

### PR DESCRIPTION
**Pull request by axetroy**
Almost complete rewrite including:

- Restructure the code
- Remove unnecessary configuration
  - `deno.packageManager`: The project does not depend on node.js, this configuration item is redundant
- Remove unnecessary commands
- Remove unnecessary files
- Remove `deno.fmtOnSave` configuration
  - This should be controlled by the built-in configuration `editor.formatOnSave`
- Remove output panel
- Fix bugs.
  - close #61
  - close #65
- Use `Github Action` as CI instead of `travis`
- No more `.denoignore` / `.denorc`
  - They are not officially supported
- Improve `deno formatter`
- Remove default configuration
	```
	"[typescript]": {
	   "editor.defaultFormatter": "justjavac.vscode-deno"
	 },
     "[markdown]": {
        "editor.defaultFormatter": "justjavac.vscode-deno"
      }
	```
- Automatically update `node_modules/typescript-deno-plugin/lib/lib.deno_runtime.d.ts` via current deno version
- Update npm dependencies
- Automatically publish to vscode Marketplace and Github when release a new version
  - Need to set secret in the project settings
    - **GH_TOKEN**: Github Personal Access Token
    - **ADO_TOKEN**: Azure DevOps Personal Access Token

### CI Status

> CI is not shown here because travis has been removed

see https://github.com/axetroy/vscode-deno/tree/refactor

截屏2020-01-2222 50 10

### Known issue

1. Can not add CI status badge in README

Can not add CI status badge in README. ref: https://github.com/Microsoft/vscode-vsce/issues/183

> Because the Github Action badge is an svg file. It will cause vsce packaging to fail

2. Need to update **ADO_TOKEN** regularly

Azure DevOps Personal Access Token is valid for a maximum of one year, so it will expire after one year.

so, It needs regular updates

### Before merge

Make sure you have done the following before merging

1. make sure CI all green except the job that release of a new version
2. add **GH_TOKEN** secret in project settings
3. add **ADO_TOKEN** secret in project settings

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -116,11 +77,21 @@
     ],
     "configurationDefaults": {
       "[typescript]": {
-        "editor.formatOnSave": false,
+        "editor.defaultFormatter": "justjavac.vscode-deno"
+      },
+      "[typescriptreact]": {
```
Comment (axetroy) [inline package.json]: Controversial here

Do we need to set the default Formatter?

File: src/utils.ts [hunk 2]
```diff
@@ -1,128 +1,29 @@
...
-    if (stderr) {
-      return;
-    }
-
-    fs.writeFileSync(path.resolve(denoDir, "lib.deno_runtime.d.ts"), stdout);
-  } catch {
-    return;
-  }
+export async function isFilepathExist(filepath: string): Promise<boolean> {
+  return promisify(stat)(filepath)
```
Comment (brandonkal) [inline src/utils.ts:22]: No need for promisify. Just use
https://nodejs.org/docs/latest-v12.x/api/fs.html#fs_fspromises_stat_path_options which is available in Electron's Node.

File: README.md [hunk 3]
```diff
@@ -44,12 +43,8 @@ You can configure the Deno extension using a `tsconfig` as described [here](http
 
 - `deno.enabled` - Enable/disable this extension. Default is `true`.
```
Comment (brandonkal) [inline README.md:45]: Remove excess whitespace

File: package.json [hunk 4]
```diff
@@ -146,11 +97,11 @@
     "@types/which": "^1.3.1",
     "husky": "^3.0.4",
     "lint-staged": "^9.2.5",
-    "typescript": "^3.7.2",
+    "typescript": "^3.7.5",
```
Comment (brandonkal) [inline package.json:100]: Current VSCode isn't at 3.7.5 yet.

File: src/extension.ts [hunk 5]
```diff
@@ -9,45 +9,38 @@ import {
...
   WorkspaceConfiguration,
-  Uri,
   languages,
   TextDocument,
   Range,
   TextEdit
 } from "vscode";
 import * as path from "path";
+import { readFile, writeFile } from "fs";
+import { promisify } from "util";
```
Comment (brandonkal) [inline src/extension.ts:19]: Prefer native fs promises API

**Review by brandonkal — CHANGES_REQUESTED** [review]

**Comment by brandonkal** [comment]
@axetroy In addition to what I've mentioned, the ISSUE_TEMPLATE directory is failing the defined `yarn run check` step.

**Comment by axetroy** [comment]
I have built a new version for this, download it here
https://github.com/axetroy/vscode-deno/releases

**Comment by David-Else** [comment]
@axetroy If there are going to be week long delays in getting pull requests merged on this repo why not start a new VS Code Market place extension? As this is a total re-write it seems logical to re-start it at 1.0?

It would make sense for all the deno contributors to get together and decide which is the 'official' version and link that on the main site. Right now the download stats for different extensions are:

https://marketplace.visualstudio.com/search?term=deno&target=VSCode&category=All%20categories&sortBy=Relevance

3.2k for this one vs 916 and 440 for the others.When Deno 1.0 comes out whatever extension is linked on the main page would soon become the predominant one. It makes no sense to have different competing versions.

I think Deno's adoption will be very much linked to the VS Code extension. Nearly everyone uses VS Code these days, once Deno starts becoming successful then any problems with the extension will stop people using Deno. Ideally it would have more than one person with the authority to update it.

**Massive thanks** to @justjavac for all the work he has done!!

**Comment by axetroy** [comment]
@David-Else 

ham. I actually have permission to merge this PR. And release a new version

But no permission to publish to vscode market.

I did plan to release another extension before and maintained by [denolib](https://github.com/denolib)

Integrate typescript-vscode-plugin together in the extension for easy maintenance

I am trying to do this.

**Comment by brandonkal** [comment]
I'm for that. Currently using a locally built version with this PR plus my PR on the TS extension. It still makes sense to publish the TS package as its own npm package but it would be cool to have them live in the same repo. See my PR on the other repo, it fixes a problem with other extensions. 

Haven't found a solution to the other bug:
`import a from "http://a.com/?broken.yaml.js"` freezes the extension. Deno doesn't like the URL either (bug) so I've changed the server to not use a query param but it would still be nice to see that not freeze the editor.

**Comment by axetroy** [comment]
I have publish a new extension to vscode market https://marketplace.visualstudio.com/items?itemName=axetroy.vscode-deno

I will try to solve problems in the new repo.

This PR is no longer updated. thanks you all.

close.

**Pull request closed without merging**